execute/group_lookup.go

```diff
@@ -7,102 +7,242 @@ import (
 )
 
 type GroupLookup struct {
-	groups groupEntries
+	// groups contains groups of group keys in sorted order.
+	// These are optimized for appending access.
+	groups []*keyGroup
 
-	//  range state
-	rangeIdx int
+	// lastIndex contains the last group that an entry was
+	// found in or appended to. This is used to optimize appending.
+	lastIndex int
 
-	// Indicates whether the group needs to be sorted.
-	needSort bool
+	// nextID is the next id that will be assigned to a key group.
+	nextID int
 }
 
-type groupEntry struct {
-	key   flux.GroupKey
-	value interface{}
+// keyGroup is a group of keys in sorted order.
+type keyGroup struct {
+	id      int // unique id for the key group within the group lookup
+	entries []keyGroupEntry
+	deleted int
 }
 
-func NewGroupLookup() *GroupLookup {
-	return &GroupLookup{
-		groups: make(groupEntries, 0, 100),
+type keyGroupEntry struct {
+	key     flux.GroupKey
+	value   interface{}
+	deleted bool
+}
+
+func (kg *keyGroup) First() flux.GroupKey {
+	return kg.entries[0].key
+}
+
+func (kg *keyGroup) Last() flux.GroupKey {
+	return kg.entries[len(kg.entries)-1].key
+}
+
+func (kg *keyGroup) set(i int, value interface{}) {
+	if kg.entries[i].deleted {
+		kg.entries[i].deleted = false
+		kg.deleted--
 	}
+	kg.entries[i].value = value
+}
+
+func (kg *keyGroup) delete(i int) {
+	kg.entries[i].value = nil
+	kg.entries[i].deleted = true
+	kg.deleted++
 }
 
-func (l *GroupLookup) findIdx(key flux.GroupKey) int {
-	if l.needSort {
-		sort.Sort(l.groups)
-		l.needSort = false
+// Index determines the location of this key within the key group.
+// It returns -1 if this key does not exist within the group.
+// It will return -1 if the entry is present, but deleted.
+func (kg *keyGroup) Index(key flux.GroupKey) int {
```

Comment:
Seems like this could be a non-exported method?

Reply:
It's technically not exported because it's on an unexported type. I like to do this because I find it easier to read and it doesn't really matter.